gli/utils.py

```diff
@@ -264,24 +260,78 @@ def sparse_to_torch(sparse_array: sp.spmatrix,
             raise TypeError(f"Unsupported sparse type {sparse_type}")
 
 
+def _find_data_files_from_json_files(data_dir):
+    """Traverse json files under dataset path and find dependent data files."""
+    json_files = []
+    for file in os.listdir(data_dir):
+        if file.endswith(".json") and file != "urls.json":
+            json_files.append(file)
+
+    def _find_data_files_helper(data):
+        """Get all entries in data with key `file` at all levels."""
+        data_files = []
+        if isinstance(data, dict):
+            for key, value in data.items():
+                if key == "file":
+                    data_files.append(value)
+                else:
+                    data_files.extend(_find_data_files_helper(value))
+        elif isinstance(data, list):
+            for item in data:
+                data_files.extend(_find_data_files_helper(item))
+        return data_files
+
+    # Get all dependent data files from json files.
+    data_files = []
+    for json_file in json_files:
+        with open(os.path.join(data_dir, json_file), "r",
+                  encoding="utf-8") as f:
+            data = json.load(f)
+            data_files.extend(_find_data_files_helper(data))
+    # Deduplicate data_files.
+    data_files = list(set(data_files))
+
+    return data_files
+
+
 def download_data(dataset: str, verbose=False):
     """Download dependent data of a configuration (metadata/task) file.
 
     Args:
         dataset (str): Name of dataset.
-        filename (str): Name of configuration file. e.g., `metadata.json`.
         verbose (bool, optional): Defaults to False.
     """
     data_dir = os.path.join(get_local_data_dir(), dataset)
-    if os.path.isdir(data_dir):
-        url_file = os.path.join(data_dir, "urls.json")
-    else:
+    if not os.path.isdir(data_dir):
         raise FileNotFoundError(f"cannot find dataset {dataset}.")
-    if not os.path.exists(url_file):
-        raise FileNotFoundError(f"cannot find url files of {dataset}.")
-    with open(url_file, "r", encoding="utf-8") as fp:
-        url_dict = json.load(fp)
-    for data_file_name, url in url_dict.items():
+
+    # Get all required dependent data files from json files.
+    data_files = _find_data_files_from_json_files(data_dir)
+
+    # First, check the local urls.json file.
+    url_file = os.path.join(data_dir, "urls.json")
+    if os.path.exists(url_file):
+        with open(url_file, "r", encoding="utf-8") as fp:
+            url_dict = json.load(fp)
+    else:
+        # Second, try to download and check the global_urls.json file.
+        global_url_file = os.path.join(data_dir, "global_urls.json")
+        _download(GLOBAL_FILE_URL, global_url_file, verbose=verbose)
```

Comment:
Are we going to download global url files every time we call this function?

Reply:
The problem is we don't know whether a previously downloaded file is up-to-date. Especially when someone is in the development stage and is testing the dataloader for files just uploaded. 

Later on, we will give a higher priority on the http request than this global file. So it might be Ok?

An improvement I can make though is to check the existence of npz files before trying to download the global file.

Reply:
Added the check

Reply:
Do you have any suggestions? @xingjian-zhang

Reply:
I see. Then I think it is okay if we are going to use an API as the primary manner to retrieve URLs.